Maple2.File.Parser/Nif/NifDocument.cs

```diff
@@ -0,0 +1,243 @@
+﻿using System.Text;
+
+namespace Maple2.File.Parser.Nif;
+
+public class NifDocument {
+    private readonly byte[] fileData;
+    private NifDocumentHeader header;
+    public EndianReader Reader { get; private set; }
+    public string RelPath { get; init; }
+    private NifBlock?[] blocks;
+    public NifBlock? ReadingBlock { get; private set; }
+
+    public List<NiPhysXProp> PhysXProps { get; init; }
+    public string VersionString { get => header.HeaderString; }
+
+    public NifDocument(string relpath, byte[] fileData) {
+        PhysXProps = new List<NiPhysXProp>();
+        this.fileData = fileData;
+        Reader = new EndianReader(Array.Empty<byte>(), false);
+        blocks = Array.Empty<NifBlock?>();
+        header = new NifDocumentHeader();
+        RelPath = relpath;
+    }
+
+    private int ReadHeaderString() {
+        int index = 0;
+        int headerStringLength = 0;
+
+        while (headerStringLength < fileData.Length && fileData[headerStringLength] != 0x0A) {
+            headerStringLength++;
+        }
+
+        header.HeaderString = Encoding.UTF8.GetString(fileData, 0, headerStringLength);
+
+        index += headerStringLength + 1;
+
+        int versionStart = headerStringLength;
+
+        while (versionStart - 1 > 0 && (fileData[versionStart - 1] == '.' || (fileData[versionStart - 1] >= '0' && fileData[versionStart - 1] <= '9')))
+            --versionStart;
+
+        for (int i = 0; i < 4; ++i) {
+            int versionEnd = versionStart;
+
+            while (versionEnd < headerStringLength && fileData[versionEnd] != '.')
+                ++versionEnd;
+
+            string versionText = Encoding.UTF8.GetString(fileData, versionStart, versionEnd - versionStart);
+
+            header.Version[i] = ushort.Parse(versionText);
+
+            versionStart = versionEnd + 1;
+        }
+
+        return index;
+    }
+
+    public NifBlock? GetBlock(int index) {
+        if (index == -1) {
+            return null;
+        }
+
+        if (blocks[index] is not null) {
+            return blocks[index];
+        }
+
+        NifBlock block;
+        string blockType = header.BlockTypes[header.BlockTypeIndices[index]];
+
+        switch (blockType) {
```

Comment:
you should use a switch expression here, surprise the IDE doesn't generate it for you